python-package/lightgbm/basic.py

```diff
@@ -15,9 +17,64 @@
 from .libpath import find_lib_path
 
 
+class _DummyLogger:
+    def info(self, msg):
+        print(msg)
+
+    def warning(self, msg):
+        warnings.warn(msg, stacklevel=3)
+
+
+_LOGGER = _DummyLogger()
+
+
+def register_logger(logger):
+    """Register custom logger.
+
+    Parameters
+    ----------
+    logger : logging.Logger
+        Custom logger.
+    """
+    if not isinstance(logger, Logger):
+        raise TypeError("Logger should inherit logging.Logger class")
+    global _LOGGER
+    _LOGGER = logger
+
+
+def _normalize_native_string(func):
+    """Join log messages from native library which come by chunks."""
+    msg_normalized = []
+
+    @wraps(func)
+    def wrapper(msg):
+        nonlocal msg_normalized
+        if msg.strip() == '':
+            msg = ''.join(msg_normalized)
+            msg_normalized = []
+            return func(msg)
+        else:
+            msg_normalized.append(msg)
+
+    return wrapper
+
+
+def _log(msg, is_warning=False):
+    """Output LightGBM's logs."""
+    if is_warning:
+        _LOGGER.warning(msg)
+    else:
+        _LOGGER.info(msg)
```

Comment:
instead of using a flag for the log level, I'd prefer functions `_log_warn()` and `_log_info()`. I think that makes the code easier to understand, and is easier to extend to other log levels (error, fatal, and debug). That is how most loggers are structured.

Reply:
I don't think that adding new arguments like `is_debug`, `is_fatal` is harder and less clear than writing new functions. I believe it's more a matter of taste. But I don't have any strong preferences, so I'm going to implement your suggestion.

Reply:
@jameslamb 

> yes I think this can and should be used by the code in `dask.py` as well.

Done in 48c405828db7071afabf43f1e68a789d5575b10d.

> instead of using a flag for the log level, I'd prefer functions `_log_warn()` and `_log_info()`

Done in 7247790078184646271e61b3bfe1d762bb57b985.